openedx/core/djangoapps/content_tagging/rest_api/v1/objecttag_export_helpers.py

```diff
@@ -38,51 +42,136 @@ def iterate_with_level(
+        object_tags=object_tag_cache.get(block_id, {}),
         children=None,
     )
 
-    blocks = [(tagged_course, course)]
+    return tagged_block, block.children if block.has_children else []
+
+
+def get_library_block_tagged_object(
+    library_block: LibraryXBlockMetadata, object_tag_cache: ObjectTagByObjectIdDict
+) -> tuple[TaggedContent, None]:
+    """
+    Returns a TaggedContent with library content block metadata and its tags,
+    and 'None' as children.
+    """
+    block_id = str(library_block.usage_key)
+    tagged_library_block = TaggedContent(
+        display_name=library_block.display_name,
+        block_id=block_id,
+        category=library_block.usage_key.block_type,
+        object_tags=object_tag_cache.get(block_id, {}),
+        children=None,
+    )
+
+    return tagged_library_block, None
+
+
+def build_object_tree_with_objecttags(
+    content_key: LibraryLocatorV2 | CourseKey,
+    object_tag_cache: ObjectTagByObjectIdDict,
+) -> TaggedContent:
+    """
+    Returns the object with the tags associated with it.
+    """
+    if isinstance(content_key, CourseKey):
+        tagged_content, children = get_course_tagged_object_and_children(
+            content_key, object_tag_cache
+        )
+    elif isinstance(content_key, LibraryLocatorV2):
+        tagged_content, children = get_library_tagged_object_and_children(
+            content_key, object_tag_cache
+        )
+    else:
+        raise NotImplementedError(f"Invalid content_key: {type(content_key)} -> {content_key}")
+
+    blocks: list[tuple[TaggedContent, list | None]] = [(tagged_content, children)]
+
+    store = modulestore()
 
     while blocks:
-        tagged_block, xblock = blocks.pop()
+        tagged_block, block_children = blocks.pop()
         tagged_block.children = []
 
-        if xblock.has_children:
-            for child_id in xblock.children:
-                child_block = store.get_item(child_id)
-                tagged_child = TaggedContent(
-                    display_name=child_block.display_name_with_default,
-                    block_id=str(child_id),
-                    category=child_block.category,
-                    object_tags=object_tag_cache.get(str(child_id), {}),
-                    children=None,
+        if not block_children:
+            continue
+
+        for child in block_children:
+            child_children: list | None
+
+            if isinstance(child, UsageKey):
+                tagged_child, child_children = get_xblock_tagged_object_and_children(
+                    child, object_tag_cache, store
+                )
+            elif isinstance(child, LibraryXBlockMetadata):
+                tagged_child, child_children = get_library_block_tagged_object(
+                    child, object_tag_cache
                 )
-                tagged_block.children.append(tagged_child)
+            else:
+                raise NotImplementedError(f"Invalid child: {type(child)} -> {child}")
```

Comment:
Now we can use the `get_tagged_children` variable set in my suggestion above:

```suggestion
            tagged_child, child_children = get_tagged_children(child, object_tag_cache)
```

Reply:
Done: 610c01aa3c6e196e1c987d7106cf9db2655f6ff9

At first, I thought we would still need a type guard before calling the function, but the types were inferred just right.

The code is way cleaner now! Thank you!

Reply:
@rpenido sorry, I don't know a better way to do this.. but I agree, having an ugly type definition is better than disabling type checking.